wsclient/src/main/java/org/hpccsystems/ws/client/platform/PhysicalMachine.java

```diff
@@ -0,0 +1,212 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+package org.hpccsystems.ws.client.platform;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+
+import org.apache.log4j.Logger;
+import org.hpccsystems.ws.client.HPCCFileSprayClient;
+import org.hpccsystems.ws.client.gen.filespray.v1_16.PhysicalFileStruct;
+import org.hpccsystems.ws.client.gen.wstopology.v1_27.TpMachine;
+import org.hpccsystems.ws.client.utils.DataSingleton;
+import org.hpccsystems.ws.client.utils.EqualsUtil;
+import org.hpccsystems.ws.client.utils.HashCodeUtil;
+import org.hpccsystems.ws.client.utils.Utils.HPCCEnvOSCode;
+
+public class PhysicalMachine extends DataSingleton
+{
+    private static final Logger       log          = Logger.getLogger(PhysicalMachine.class.getName());
+    private static Map<Integer, PhysicalMachine> physicalMachines = new HashMap<Integer, PhysicalMachine>();
+
+    public static synchronized PhysicalMachine get(Platform platform, String name)
+    {
+        PhysicalMachine physicalMachine = new PhysicalMachine(platform, name);
+        if (physicalMachines.containsKey(physicalMachine.hashCode()))
+        {
+            return physicalMachines.get(physicalMachine.hashCode());
+        }
+        else
+        {
+            physicalMachines.put(physicalMachine.hashCode(), physicalMachine);
+        }
+        return physicalMachine;
```

Comment:
Should we be calling refresh on the PhysicalMachine before returning it? The monitorThread in DataSingleton will call refresh eventually, but until then it will be in an invalid state.

Reply:
The physicalmachine file list could indeed be outdated, but the refresh can be delayed until the caller requests the file list, potentially avoiding an unecessary wstopology call